core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1966,29 +1966,79 @@ private[zio] trait ZIOFunctions extends Serializable {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    def arbiter(
+      fibers: Iterable[Fiber[E, _]],
+      promise: Promise[E, List[B]],
+      buffer: Array[B],
+      todo: Ref[Int],
+      idx: Int
+    )(res: Exit[E, B]): ZIO[R, Nothing, Unit] =
+      res.foldM[R, Nothing, Unit](
+        e => promise.halt(e).unit *> Fiber.interruptAll(fibers),
+        a =>
+          todo.modify { t =>
+            buffer.update(idx, a)
+            if (t == 1) promise.succeed(buffer.toList).unit -> 0 else UIO.unit -> (t - 1)
+          }.flatten
+      )
+
+    (for {
+      size    <- UIO.effectTotal(as.size)
+      todo    <- Ref.make(size)
+      buffer  <- UIO.effectTotal(new Array[Any](size).asInstanceOf[Array[B]])
```

Comment:
Will this not cause a java.lang.ClassCastException ? A quick experiment from the scala repl:

```
scala> new Array[Any](100).asInstanceOf[Array[String]]
java.lang.ClassCastException: [Ljava.lang.Object; cannot be cast to [Ljava.lang.String;
  ... 36 elided
```

Reply:
I think the trick here is not to cast it to specific type, but only to a generic one.

Reply:
I still don't get how this works at runtime though, because the array will need to get a concrete type at some stage right?

```
scala> def test[A](a:A):Array[A] = new Array[Any](1).asInstanceOf[Array[A]]
test: [A](a: A)Array[A]

scala> test(1)
java.lang.ClassCastException: [Ljava.lang.Object; cannot be cast to [I
  ... 28 elided
```

Reply:
Isn't is caused by specialization? There is even a test that uses `foreachPar` on a `List[Int]`: https://github.com/hanny24/scalaz-zio/blob/35bd04f28cedeb3c988a83af206c37d48d3b58ad/core-tests/shared/src/test/scala/zio/ZIOSpec.scala#L70

Reply:
Yeah saw that test now. I have to admit that I don't understand how this works. It clearly does though :)